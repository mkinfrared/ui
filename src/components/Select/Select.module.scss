@import "../../styles/common.scss";

.Select {
  --border-color: var(--black);
  --background: transparent;
  --icon-color: var(--black);
  --left-spacing: 0;
  --right-spacing: 0;

  display: block;
  width: 100%;

  .label {
    cursor: pointer;
    margin-bottom: 0.6rem;
    text-transform: capitalize;
  }

  .container {
    border: 1px solid;
    border-color: var(--border-color);
    border-radius: 1rem;
    font-family: Roboto, sans-serif;
    font-size: var(--text-size);
    font-style: normal;
    font-weight: 500;
    line-height: 1.8;
    overflow: hidden;
    position: relative;
    width: 100%;

    :where(&) {
      @include theme("dark") {
        @include dark;

        --input-color: var(--white);
      }

      @include theme("light") {
        @include light;
      }
    }

    .arrow {
      transform: rotateX(0deg) rotateZ(90deg);
      transition: transform 250ms var(--ease-1);
    }

    &[aria-expanded="true"] {
      .arrow {
        transform: rotateX(180deg) rotateZ(90deg);
      }
    }

    :is(textarea, input) {
      background-color: transparent;
      border: none;
      caret-color: var(--orange);
      color: inherit;
      font-size: var(--text-size);
      font-weight: 400;
      height: 100%;
      line-height: 1;
      overflow: hidden;
      padding: 1.2rem 1.4rem;
      resize: none;
      width: 100%;

      &:-webkit-autofill,
      &:-webkit-autofill:active,
      &:-webkit-autofill:hover,
      &:-webkit-autofill:focus {
        background-color: var(--background);
        border: none;

        // it's not possible to use transparent background color
        // to override chrome autofill styles, so background needs
        // to be set to a certain color, to keep the initial background
        // transparent the transition delay can be used
        box-shadow: 0 0 0 1000px var(--background-color) inset;
        caret-color: var(--orange);
        font-size: 1.4rem;
        -webkit-text-fill-color: var(--black);
        transition: background-color 7200s ease-in-out 0s; // 2 hr delay
      }

      &:-webkit-autofill::selection,
      &::selection {
        background-color: var(--orange);
        color: var(--white);
      }

      &::placeholder {
        color: var(--grey);
      }

      &:focus-visible {
        outline: none;
      }

      svg {
        transition: transform 250ms ease-in-out;
      }

      &[aria-expanded="true"] svg {
        transform: rotateX(180deg);
      }
    }

    .prefix,
    .suffix {
      aspect-ratio: 1/1;
      color: var(--icon-color);
      height: 1.4rem;
      line-height: 0;
      max-width: 1.4rem;
      position: absolute;
      text-align: center;
      top: 50%;
      transform: translateY(-50%);

      svg {
        fill: currentcolor;
        max-height: 100%;
        max-width: 100%;
      }
    }

    .prefix {
      inset-inline-start: 1.2rem;
    }

    .suffix {
      cursor: pointer;
      inset-inline-end: 1.2rem;
    }
  }

  .errorContainer {
    height: 1.8rem;
    margin-top: 0.6rem;

    .error {
      color: var(--error);
      line-height: 1.8rem;
      opacity: 1;
      overflow: hidden;
      padding: 0;
      text-overflow: ellipsis;
      transform: scaleY(1);
      transform-origin: top;
      transition: transform 250ms var(--ease-4), opacity 250ms var(--ease-4);
      white-space: nowrap;

      &:empty {
        opacity: 0;
        transform: scaleY(0);
      }
    }
  }

  &:has(.error:not(:empty)) {
    --border-color: var(--error);
  }

  &:has(input:disabled) {
    --background: var(--grey);

    .label {
      cursor: default;
    }
  }

  &:has(.prefix) {
    --left-spacing: 2.8rem;
  }

  &:has(.suffix) {
    --right-spacing: 2.8rem;
  }

  .container:focus-visible {
    @include focus-visible;
  }
}
