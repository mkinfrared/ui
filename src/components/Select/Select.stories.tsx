import { Meta, Story } from "@storybook/react";
import { useState } from "react";

import Select from "./Select";
import { SelectProps } from "./Select.type";
import css from "./Story.module.scss";

export default {
  title: "UI/Select",
  component: Select,
  parameters: {
    componentSubtitle: "Subtitle goes here",
  },
} as Meta;

const Template: Story<SelectProps<string>> = (args) => {
  const [selectedOption, setSelectedOption] = useState<string>();

  const states = [
    "Alabama",
    "Alaska",
    "Arizona",
    "Arkansas",
    "California",
    "Colorado",
    "Connecticut",
    "Delaware",
    "Florida",
    "Georgia",
    "Hawaii",
    "Idaho",
    "Illinois",
    "Indiana",
    "Iowa",
    "Kansas",
    "Kentucky",
    "Louisiana",
    "Maine",
    "Maryland",
    "Massachusetts",
    "Michigan",
    "Minnesota",
    "Mississippi",
    "Missouri",
    "Montana",
    "Nebraska",
    "Nevada",
    "New Hampshire",
    "New Jersey",
    "New Mexico",
    "New York",
    "North Carolina",
    "North Dakota",
    "Ohio",
    "Oklahoma",
    "Oregon",
    "Pennsylvania",
    "Rhode Island",
    "South Carolina",
    "South Dakota",
    "Tennessee",
    "Texas",
    "Utah",
    "Vermont",
    "Virginia",
    "Washington",
    "West Virginia",
    "Wisconsin",
    "Wyoming",
  ];

  const getOptionLabel = (value?: string) => value ?? "";
  const getActiveOption = (value: string) => value === selectedOption;

  return (
    <div className={css.Story}>
      <Select
        {...args}
        options={states}
        getOptionLabel={getOptionLabel}
        getActiveOption={getActiveOption}
        onChange={setSelectedOption}
        value={selectedOption}
      />
    </div>
  );
};

const Default = Template.bind({});

Default.parameters = {
  docs: {
    storyDescription: "Story description",
  },
};

export { Default };
