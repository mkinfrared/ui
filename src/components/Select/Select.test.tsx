import { fireEvent, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import Select from "./Select";

describe("<Select />", () => {
  let active: number | undefined = 44;

  const options = [42, 43, 44, 45];
  const getActiveOption = (value: number) => value === active;
  const getOptionLabel = (value?: number) => `marklar ${value}`;
  const onChange = jest.fn();

  const Component = (
    <Select
      getActiveOption={getActiveOption}
      getOptionLabel={getOptionLabel}
      onChange={onChange}
      options={options}
    />
  );

  afterEach(() => {
    jest.resetAllMocks();

    active = 44;
  });

  it("should be defined", () => {
    expect(Select).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should contain a data test id", () => {
    const { getByTestId } = render(Component);
    const element = getByTestId("Select");

    expect(element).toBeDefined();
  });

  it("should open listbox on 'ArrowDown' key press", async () => {
    const { queryByTestId, getByRole } = render(Component);
    const combobox = getByRole("combobox");

    let listbox = queryByTestId("Menu");

    expect(listbox).toBe(null);

    await userEvent.tab();

    fireEvent.keyDown(combobox, { key: "ArrowDown" });

    listbox = queryByTestId("Menu");

    expect(listbox).not.toBe(null);
  });

  it("should focus the first element on the list if none is active on 'ArrowDown' key press", async () => {
    active = undefined;

    const { queryByTestId, getByRole, getByText } = render(Component);
    const combobox = getByRole("combobox");

    let listbox = queryByTestId("Menu");

    expect(listbox).toBe(null);

    await userEvent.tab();

    fireEvent.keyDown(combobox, { key: "ArrowDown" });

    listbox = queryByTestId("Menu");

    expect(listbox).not.toBe(null);

    fireEvent.keyDown(combobox, { key: "ArrowDown" });

    const menuItem = getByText("marklar 42");

    expect(menuItem).toHaveFocus();
  });

  it("should focus the active element on 'ArrowDown' key press", async () => {
    const { queryByTestId, getByRole, getByText } = render(Component);
    const combobox = getByRole("combobox");

    let listbox = queryByTestId("Menu");

    expect(listbox).toBe(null);

    await userEvent.tab();

    fireEvent.keyDown(combobox, { key: "ArrowDown" });

    listbox = queryByTestId("Menu");

    expect(listbox).not.toBe(null);

    fireEvent.keyDown(combobox, { key: "ArrowDown" });

    const menuItem = getByText("marklar 44");

    expect(menuItem).toHaveFocus();
  });

  it("should focus the combobox on 'Enter' press if the listbox is opened", async () => {
    const { queryByTestId, getByRole, getByText } = render(Component);
    const combobox = getByRole("combobox");

    let listbox = queryByTestId("Menu");

    expect(listbox).toBe(null);

    await userEvent.tab();

    fireEvent.keyDown(combobox, { key: "ArrowDown" });

    listbox = queryByTestId("Menu");

    expect(listbox).not.toBe(null);

    fireEvent.keyDown(combobox, { key: "ArrowDown" });

    const menuItem = getByText("marklar 44");

    expect(menuItem).toHaveFocus();

    fireEvent.keyDown(combobox, { key: "Enter" });

    expect(combobox).toHaveFocus();
  });

  it("should focus the input on 'ArrowRight' press", async () => {
    const { queryByTestId, getByRole, getByText } = render(Component);
    const combobox = getByRole("combobox");
    const input = combobox.querySelector("input");

    let listbox = queryByTestId("Menu");

    expect(listbox).toBe(null);

    await userEvent.tab();

    fireEvent.keyDown(combobox, { key: "ArrowDown" });

    listbox = queryByTestId("Menu");

    expect(listbox).not.toBe(null);

    fireEvent.keyDown(combobox, { key: "ArrowDown" });

    const menuItem = getByText("marklar 44");

    expect(menuItem).toHaveFocus();

    fireEvent.keyDown(combobox, { key: "ArrowRight" });

    expect(input).toHaveFocus();
  });

  it("should call 'onChange' from props on MenuItem click and close listbox", async () => {
    const { queryByTestId, getByRole, getByText } = render(Component);
    const combobox = getByRole("combobox");

    let listbox = queryByTestId("Menu");

    expect(listbox).toBe(null);

    await userEvent.tab();

    fireEvent.keyDown(combobox, { key: "ArrowDown" });

    listbox = queryByTestId("Menu");

    expect(listbox).not.toBe(null);

    const menuItem = getByText("marklar 43");

    fireEvent.click(menuItem);

    listbox = queryByTestId("Menu");

    expect(onChange).toHaveBeenCalledTimes(1);

    expect(onChange).toHaveBeenCalledWith(43);

    expect(listbox).toBe(null);
  });

  it("should close listbox on 'Escape' press", async () => {
    const { queryByTestId, getByRole } = render(Component);
    const combobox = getByRole("combobox");

    let listbox = queryByTestId("Menu");

    expect(listbox).toBe(null);

    await userEvent.tab();

    fireEvent.keyDown(combobox, { key: "ArrowDown" });

    listbox = queryByTestId("Menu");

    expect(listbox).not.toBe(null);

    fireEvent.keyDown(combobox, { key: "Escape" });

    listbox = queryByTestId("Menu");

    expect(listbox).toBe(null);
  });
});
