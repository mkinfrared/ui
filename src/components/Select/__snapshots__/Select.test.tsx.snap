// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should match the snapshot 1`] = `
<div>
  <label
    class="Select"
    data-testid="Select"
  >
    <div
      aria-controls="listbox1"
      aria-expanded="false"
      class="container"
      role="combobox"
      tabindex="0"
    >
      <input
        type="text"
        value="marklar undefined"
      />
      <div
        class="suffix"
      >
        <svg
          class="arrow"
        >
          chevron.svg
        </svg>
      </div>
    </div>
    <div
      class="errorContainer"
    >
      <p
        class="Text error"
        data-testid="Text"
      />
    </div>
  </label>
</div>
`;
